packages/dds/task-manager/src/oldestClientObserver.ts

```diff
@@ -0,0 +1,115 @@
+ *
+ * It is still experimental and under development.  Please do try it out, but expect breaking changes in the future.
+ *
+ * @remarks
+ * ### Creation
+ *
+ * The `OldestClientObserver` constructor takes an `IContainerRuntime`:
+ *
+ * ```typescript
+ * const oldestClientObserver = new OldestClientObserver(containerRuntime);
+ * ```
+ *
+ * ### Usage
+ *
+ * To check if the local client is the oldest, use the `isOldest()` method.
+ *
+ * ```typescript
+ * if (oldestClientObserver.isOldest()) {
+ *     console.log("I'm the oldest");
+ * } else {
+ *     console.log("Someone else is older");
+ * }
+ * ```
+ *
+ * ### Eventing
+ *
+ * `OldestClientObserver` is an `EventEmitter`, and will emit events when the local client becomes the oldest and when
+ * it is no longer the oldest.
+ *
+ * ```typescript
+ * oldestClientObserver.on("becameOldest", () => {
+ *     console.log("I'm the oldest now");
+ * });
+ *
+ * oldestClientObserver.on("lostOldest", () => {
+ *     console.log("I'm not the oldest anymore");
+ * });
+ * ```
+ */
+export class OldestClientObserver extends EventEmitter implements IOldestClientObserver {
+    private readonly quorum: IQuorum;
+    private currentIsOldest: boolean = false;
+    constructor(private readonly containerRuntime: IContainerRuntime) {
+        super();
+        this.quorum = this.containerRuntime.getQuorum();
+        this.currentIsOldest = this.computeIsOldest();
+        this.quorum.on("addMember", this.updateOldest);
+        this.quorum.on("removeMember", this.updateOldest);
+        containerRuntime.on("connected", this.updateOldest);
+        containerRuntime.on("disconnected", this.updateOldest);
+    }
+
+    public isOldest(): boolean {
+        return this.currentIsOldest;
+    }
+
+    private readonly updateOldest = () => {
+        const oldest = this.computeIsOldest();
+        if (this.currentIsOldest !== oldest) {
+            this.currentIsOldest = oldest;
+            if (oldest) {
+                this.emit("becameOldest");
+            } else {
+                this.emit("lostOldest");
+            }
+        }
+    };
+
+    private computeIsOldest(): boolean {
+        // If the container is detached, we are the only ones that know about it and are the oldest by default.
+        if (this.containerRuntime.attachState === AttachState.Detached) {
+            return true;
+        }
+
+        // Without a clientId we can't verify if we're oldest, assume we are not
+        if (this.containerRuntime.clientId === undefined) {
+            return false;
+        }
+
+        // If we're not connected we can't be the oldest connected client
+        if (!this.containerRuntime.connected) {
+            return false;
+        }
+
+        const selfSequencedClient = this.quorum.getMember(this.containerRuntime.clientId);
+        if (selfSequencedClient === undefined) {
+            return false;
+        }
+
+        const members = this.quorum.getMembers();
```

Comment:
Please add
assert(selfSequencedClient.shouldHaveLeft !== true);

Reply:
I don't plan to add this assert for a few reasons:
1. `ConnectionStateHandler` marks `shouldHaveLeft` at the same time that it updates its `clientId`, so it seems the only way this would hit is if there are multiple `ConnectionStateHandler`s per single `Quorum` which will blow up in many other ways.
2. `ILocalSequencedClient` is currently internal to `@fluidframework/container-loader`, and I'd prefer not to export it and also prefer not to add a container-loader dependency.
3. I don't like that we're manipulating the `ISequencedClient` members of the quorum to be an unexpected type, IMO `ConnectionStateHandler` should not be mutating the quorum members.